Terminal.Gui/Input/Key.cs

```diff
@@ -72,11 +73,41 @@ public class Key : EventArgs, IEquatable<Key> {
 	public Key () : this (KeyCode.Null) { }
 
 	/// <summary>
-	///   Constructs a new <see cref="Key"/> from the provided Key value
+	/// Constructs a new <see cref="Key"/> from the provided Key value
 	/// </summary>
 	/// <param name="k">The key</param>
 	public Key (KeyCode k) => KeyCode = k;
 
+	/// <summary>
+	/// Constructs a new <see cref="Key"/> from a char.
+	/// </summary>
+	/// <remarks>
+	/// <para>
+	/// The key codes for the A..Z keys are encoded as values between 65 and 90 (<see cref="KeyCode.A"/> through <see cref="KeyCode.Z"/>).
+	/// While these are the same as the ASCII values for uppercase characters, they represent *keys*, not characters.
+	/// Therefore, this constructor will store 'A'..'Z' as <see cref="KeyCode.A"/>..<see cref="KeyCode.Z"/> with the
+	/// <see cref="KeyCode.ShiftMask"/> set and will
+	/// store `a`..`z` as <see cref="KeyCode.A"/>..<see cref="KeyCode.Z"/>.
+	/// </para>
+	/// </remarks>
+	/// <param name="ch"></param>
+	public Key (char ch)
+	{
+		switch (ch) {
+		case >= 'A' and <= 'Z':
+			// Upper case A..Z mean "Shift-char" so we need to add Shift
+			KeyCode = (KeyCode)ch | KeyCode.ShiftMask;
+			break;
+		case >= 'a' and <= 'z':
+			// Lower case a..z mean no shift, so we need to store as Key.A...Key.Z
+			KeyCode = (KeyCode)(ch - 32);
+			return;
+		default:
+			KeyCode = (KeyCode)ch;
+			break;
+		}
+	}
```

Comment:
@tig you can leverage this code https://github.com/BDisp/Terminal.Gui/blob/9349d542974c6eb59044b3ca4716cc1a25101321/Terminal.Gui/Input/Key.cs#L195-L253 to get the shift or not based on more range of characters to allow test this:

```cs
	[InlineData ('á', KeyCode.A)]
	[InlineData ('Á', (KeyCode)'Á' | KeyCode.ShiftMask)]

```

Reply:
Here is the full code https://github.com/BDisp/Terminal.Gui/blob/9349d542974c6eb59044b3ca4716cc1a25101321/Terminal.Gui/ConsoleDrivers/ConsoleKeyMapping.cs#L196-L294

Reply:
I don't think that is correct behavior. Is it  really needed for VK packet support or is it something you think is a nice-to-have?

If it's not absolutely needed for VK packet support then I'd prefer to have A..Z be the only keys supported in the test.

Reply:
Ah if it's only needed for the tests ok, but you still need this for `TextView` recognize that they are valid characters to pass on the `GetIsKeyCodeAtoZ` method to insert the text.

Reply:
An aside, as I've been looking at all these tests:

Might I suggest using MemberData to generate test cases instead of all the InlineData?

It's xUnit's analogue to NUnit's TestCaseSource and would allow you to have more exhaustive and authoritative tests with less work and less code.

The general pattern is:

```
[Theory]
[MemberData(nameof(TheoryA_Cases))]
public void TheoryA( string s, int i, object o )
{
  // Test code
}

private static IEnumerable<object[]> TheoryA_Cases()
{
  // Code that provides an IEnumerable of 3-element object arrays that satisfy the parameters of the test.
}
```

xUnit's documentation is sorely lacking, but that's the basic pattern for using that.

I do not know if xUnit has a way to generate tests from combinatorial parameter sources, but that effect can be achieved by just doing the combinatorial loops yourself in the case generator method.

Reply:
Someone did contribute an example for xUnit, of a way to provide yourself with fairly easily reusable combinatorial case generators:

https://github.com/xunit/samples.xunit/blob/main/TheoryDataExamples/MatrixTheoryData.cs

That is only for 2 parameters, but extending it to 3 or more is fairly straightforward.

Reply:
You need it because this unit test is wrong:

`[InlineData ('Ó', (KeyCode)'Ó')]`

Should be:

`[InlineData ('Ó', (KeyCode)'Ó' | KeyCode.ShiftMask)]`

Because Ó is a capital letter.

Reply:
Change to this and you'll see.

```cs
	public Key (char ch)
	{
		switch (ch) {
		case >= 'A' and <= 'Z':
			// Upper case A..Z mean "Shift-char" so we need to add Shift
			KeyCode = (KeyCode)ch | KeyCode.ShiftMask;
			break;
		case char when (ch - 32) is >= 'A' and <= 'Z':
			// Lower case a..z mean no shift, so we need to store as Key.A...Key.Z
			KeyCode = (KeyCode)(ch - 32);
			return;
		case >= 'À' and <= 'Ý':
			// Upper case À..Ý mean "Shift-char" so we need to add Shift
			KeyCode = (KeyCode)ch | KeyCode.ShiftMask;
			break;
		case char when (ch - 32) is >= 'à' and <= 'ý':
			// Lower case à..ý mean no shift, so we need to store as Key.A...Key.Z
			KeyCode = (KeyCode)(ch - 32);
			return;
		default:
			KeyCode = (KeyCode)ch;
			break;
		}
	}

```

Reply:
@dodexahedron we also have a `ClassData` unit test here https://github.com/BDisp/Terminal.Gui/blob/9349d542974c6eb59044b3ca4716cc1a25101321/UnitTests/ConsoleDrivers/ConsoleKeyMappingTests.cs#L102.

Reply:
> I don't think that is correct behavior. Is it really needed for VK packet support or is it something you think is a nice-to-have?
> 
> If it's not absolutely needed for VK packet support then I'd prefer to have A..Z be the only keys supported in the test.

Remember that you created a `Key` class that rely if the shift key is enabled or not for a certain character. So it isn't a matter if it's only needed for VK packet or not. Is how the keystroke really work. If you want to pass the char '!' in a `Key` instance you'll need to know if the shift key will be assed or not. In this case the shift is needed because you'll need to press `Shift+D1` to obtain the '!'.

Reply:
> you still need this for `TextView` recognize that they are valid characters to pass on the `GetIsKeyCodeAtoZ` method to insert the text.

Why?

Reply:
Because I added the shift key to the '!' and others characters that only can be showing by pressing the shift key. But, like the 'A', it's needed to remove the shift key from the `Key` and only send the character '!', otherwise the key is handled as key binding `(Key)'!' | Key.Shift`.

Reply:
> > I don't think that is correct behavior. Is it really needed for VK packet support or is it something you think is a nice-to-have?
> > If it's not absolutely needed for VK packet support then I'd prefer to have A..Z be the only keys supported in the test.
> 
> Remember that you created a `Key` class that rely if the shift key is enabled or not for a certain character. So it isn't a matter if it's only needed for VK packet or not. Is how the keystroke really work. If you want to pass the char '!' in a `Key` instance you'll need to know if the shift key will be assed or not. In this case the shift is needed because you'll need to press `Shift+D1` to obtain the '!'.

I disagree. The only special case required is for A...Z. All other cases work fine. I don't see any scenario where an app developer needs to know that `!` was the result of KeyCode.D1 | KeyCode.ShiftMask. All they need to know is the user meant to type '!'. 

If you have a specific use-case that requires all of the added complexity you are suggesting please describe it. 

In the current implementation, if I press KeyCode.D7 | KeyCode.ShiftMask:

- ENG Keyboard: 
<img width="274" alt="image" src="https://github.com/gui-cs/Terminal.Gui/assets/585482/8600bf3f-d40c-4d7f-99c8-ce7ee257acc4">

- DEU Keyboard:
<img width="383" alt="image" src="https://github.com/gui-cs/Terminal.Gui/assets/585482/51b73158-263d-4413-96e7-0f81aee3c371">

This is the behavior I think apps want to see.

(Note: v2_devlelop somehow broke how shift/ctrl/alt key presses on Windows worked and are showing "Null". I will fix this asap).

Reply:
> @dodexahedron we also have a `ClassData` unit test here https://github.com/BDisp/Terminal.Gui/blob/9349d542974c6eb59044b3ca4716cc1a25101321/UnitTests/ConsoleDrivers/ConsoleKeyMappingTests.cs#L102.

Right

The point is it should be used for the majority of these tests. Spot checks miss things, and this is a crucial class that warrants positive and exhaustive proof. There are also several cases of duplicated test cases, which underlines that humans writing test case inputs one by one is a highly error-prone process.

Reply:
> > > I don't think that is correct behavior. Is it really needed for VK packet support or is it something you think is a nice-to-have?
> > > If it's not absolutely needed for VK packet support then I'd prefer to have A..Z be the only keys supported in the test.
> > 
> > 
> > Remember that you created a `Key` class that rely if the shift key is enabled or not for a certain character. So it isn't a matter if it's only needed for VK packet or not. Is how the keystroke really work. If you want to pass the char '!' in a `Key` instance you'll need to know if the shift key will be assed or not. In this case the shift is needed because you'll need to press `Shift+D1` to obtain the '!'.
> 
> I disagree. The only special case required is for A...Z. All other cases work fine. I don't see any scenario where an app developer needs to know that `!` was the result of KeyCode.D1 | KeyCode.ShiftMask. All they need to know is the user meant to type '!'.
> 
> If you have a specific use-case that requires all of the added complexity you are suggesting please describe it.
> 
> In the current implementation, if I press KeyCode.D7 | KeyCode.ShiftMask:
> 
> * ENG Keyboard:
> 
> <img alt="image" width="274" src="https://private-user-images.githubusercontent.com/585482/292790626-8600bf3f-d40c-4d7f-99c8-ce7ee257acc4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDM1NTc0MzUsIm5iZiI6MTcwMzU1NzEzNSwicGF0aCI6Ii81ODU0ODIvMjkyNzkwNjI2LTg2MDBiZjNmLWQ0MGMtNGQ3Zi05OWM4LWNlN2VlMjU3YWNjNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBSVdOSllBWDRDU1ZFSDUzQSUyRjIwMjMxMjI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDIzMTIyNlQwMjE4NTVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zODFmOTQzMzcyZWI5NzIyN2I5NmUyMDUyNzc4YjhmZmEyZDExZjE4ZDE4NDMxMzE1NGFhZTIwMTM5Y2EzOTFlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.JrtMlK9bGHQ-paB7LlF-xgscpzgbChYRYQ6CKmkd45I">
> * DEU Keyboard:
> 
> <img alt="image" width="383" src="https://private-user-images.githubusercontent.com/585482/292790679-51b73158-263d-4413-96e7-0f81aee3c371.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDM1NTc0MzUsIm5iZiI6MTcwMzU1NzEzNSwicGF0aCI6Ii81ODU0ODIvMjkyNzkwNjc5LTUxYjczMTU4LTI2M2QtNDQxMy05NmU3LTBmODFhZWUzYzM3MS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBSVdOSllBWDRDU1ZFSDUzQSUyRjIwMjMxMjI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDIzMTIyNlQwMjE4NTVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mMGNlNzJiY2YxZWRiZmM4NTBhZWVkYWE0OTZlOTk4MTI0Y2ZmZjhlZmIyODlkZTQ1OTRiNGI4OWY1YWI5NGZjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.6SLUeXkh6Ua9wPjo6ZlmdJKi6HzS2RYllWgCnk58ads">
> This is the behavior I think apps want to see.
> 
> (Note: v2_devlelop somehow broke how shift/ctrl/alt key presses on Windows worked and are showing "Null". I will fix this asap).

Key binds.

How do you know if a user pressed ctrl+shift+D1 or simply pasted the string "!" if you don't communicate the keys?

Reply:
Before the creating of the `Key` class the shift key alone never was added on a character because of this confusion and it also should not be added now for A..Z. This was what starting this confusion. The shift key only must be added to `Key` if it's accomplished with another modifiers but never alone. It only have to be noted in the property `IsShift` bool.

Reply:
The only exceptions where a shift key can be added alone to a `Key` is when a non-printable character is used, like F1..F12, cursor keys, delete, backspace, etc.

Reply:
Right and that's fine, because the information is still there. As long as you can round-trip it, you're golden.

Reply:
> The only exceptions where a shift key can be added alone to a `Key` is when a non-printable character is used, like F1..F12, cursor keys, delete, backspace, etc.

Makes perfect sense.

Theoretical question, though:

Say I have a string on the clipboard that has all characters my keyboard can produce - letters, digits, symbols, and whitespace - and I paste that string while holding ctrl.

Would characters such as the ! be interpreted as "ctrl+shift+D1"? Or would they come across as just D1? Or would they come across as ctrl+D1? 🤔

Reply:
In my PR I was creating as `(KeyCode)'!' | KeyCode.ShiftMask` to leverage the character. `ConsoleKeyInfo` parses it as `KeyChar='!', Key=ConsoleKey.D1, ConsoleModifiers.Shift`.
Pasting from the clipboard only will return the clipboard content and not the ctrl key and will return as `(KeyCode)'!' | KeyCode.ShiftMask`. In `TextField` you use the `Ctrl+V` to paste and in `TextView` you use the `Ctrl+Y`.
But now the problem adding only a shift key on a character other than A..Z, because the shift key is removed from the `Key` before insert the text in a `TextView` and not for the others characters. Because of this issue, `Terminal.Gui` never added the shift key alone on printable character. Now it's only possible to A..Z and thus is impossible normalize all.

Reply:
Tell me the user scenario that no longer works.

Reply:
> I don't think that is correct behavior. Is it really needed for VK packet support or is it something you think is a nice-to-have?
> 
With the current `Key` code is the correct behavior if the `GetIsKeyCodeAtoZ` is to persist always because the shift key does not is present only from A..Z. It's good to have, although the shift key must be removed before sending to a consumer and the `IsShift` indicates whether the shift key was pressed.

> If it's not absolutely needed for VK packet support then I'd prefer to have A..Z be the only keys supported in the test.

With the current `Key` code that only is supporting the shift key for A..Z, this is needed for the  VK packet support as well.
I already proved that `'Ó'` is `(Key)'Ó' | Key.Shift` but there are also others printable character that has the shift key, like `!"#$%&/()=?»*ª>;:_`. There are two options:
1. Not including only the shift key to any printable character as before which will not provide the `IsShift` property as `true`.
2. Including the shift key to any printable character which will provide the `IsShift` property as `true`. I prefer this one but if the first is preferable the the `GetIsKeyCodeAtoZ` must be removed and treat all as equal as before.